lua/starfall/sflib.lua

```diff
@@ -495,90 +494,95 @@ SF.Parent = {
 				end,
 				function(self)
 					self.ent:SetParent()
-				end
+				end,
 			},
 			bone = {
 				function(self)
 					self.ent:FollowBone(self.parent, self.param)
 				end,
 				function(self)
 					self.ent:FollowBone(NULL, 0)
-				end
+				end,
 			}
 		},
-
+		
 		setParent = function(self, parent, type, param)
 			if self.parent and self.parent:IsValid() then
-				self.parent.sf_parent.children[self.ent] = nil
-				self:unParentType()
+				self:remove()
+				self.parent.sfParent.children[self.ent] = nil
 			end
+			
 			if parent then
-				self.parent = parent
 				self.param = param
-				self.setParentType, self.unParentType = unpack(self.setParents[type])
-
-				local sf_parent = parent.sf_parent
-				if sf_parent then
-					sf_parent.children[self.ent] = self
-				end
+				self.parent = parent
+				self.apply, self.remove = unpack(self.types[type])
+				parent.sfParent.children[self.ent] = self
+				
 				self:updateTransform()
-				self:setParentType()
+				self:apply()
 			else
-				self.parent = nil
 				self.param = nil
-				self.setParentType = nil
-				self.unParentType = nil
+				self.parent = nil
+				self.apply, self.remove = nil, nil
 			end
 		end,
-
+		
 		fix = function(self)
-			local empty = true
-			if self.parent and self.parent:IsValid() then
-				self:applyTransform()
-				self:setParentType()
-				empty = false
-			end
+			local has_children
 			for child, data in pairs(self.children) do
 				if child:IsValid() then
 					data:applyTransform()
-					data:setParentType()
-					empty = false
+					data:apply()
+					has_children = true
+					
+					if child.sfParent then
+						child.sfParent:fix()
+					end
 				else
 					self.children[child] = nil
 				end
 			end
-			if empty then
-				self.ent.sf_parent = nil
+			if not has_children then
+				self.ent.sfParent = nil
```

Comment:
This will cause the children of a parent to lose their sfParent table. If that is ok, then the `child.sfParent = setmetatable({` code should be removed

Reply:
Yeah, it's a bug good, it should not remove the `child.sfParent` unless it doesn't have a valid parent.

Just had a power outrage and it's Sunday so IDK if I'll be able to fix it anytime soon, so I checked `Allow edits by maintainers` if you wanna have a go at it.

Reply:
Fixed, now `sfParent` will only be removed if an entity doesn't have valid children as well as a valid parent.